[Bug]: The node stops working after a transaction is sent

### Version

next

### Other packages versions

miden-client: next (more specifically on commit `2f8707faeded3725a1d732310f616e3e2170a50d`)

### What happened?

The node works correctly until a transaction from the client is sent and then it "freezes", all of the requests after that don't get a response. No error message is printed in the node console.

We noticed that after reverting the commit `[chore: update dependencies (#476)]` the node didn't have this problem. More specifically if we only revert the commit [Update tonic et al](https://github.com/0xPolygonMiden/miden-node/pull/476/commits/636c6ab1065c0bacc2bc3945eec368527936989f) the node also was suddenly fixed. It might be a problem with the new versions of prost and tonic (the client in next is also using these new versions but it seems that it doesn't cause any issues).

### What should have happened?

The node should have processed the transaction and continue responding.

### How can this be reproduced?

With a node (from branch next) running locally. Run the command `make integration-test` in from the root of the `miden-client` repo (also in branch next).

### Relevant log output

```shell
The last log printed is always the same:


2024-09-06T16:35:45.727004Z  INFO store:get_block_inputs: miden-store: crates/store/src/server/api.rs:383: new
2024-09-06T16:35:45.727019Z  INFO store:get_block_inputs:validate_nullifiers: miden-store: crates/store/src/server/api.rs:591: new
2024-09-06T16:35:45.727028Z  INFO store:get_block_inputs:validate_nullifiers: miden-store: crates/store/src/server/api.rs:591: close, time.busy: 1.42µs, time.idle: 7.83µs
2024-09-06T16:35:45.727038Z  INFO store:get_block_inputs:validate_notes: miden-store: crates/store/src/server/api.rs:601: new
2024-09-06T16:35:45.727046Z  INFO store:get_block_inputs:validate_notes: miden-store: crates/store/src/server/api.rs:601: close, time.busy: 1.04µs, time.idle: 6.62µs
2024-09-06T16:35:45.727058Z  INFO store:get_block_inputs:select_block_header_by_block_num: miden-store: crates/store/src/db/mod.rs:249: new
2024-09-06T16:35:45.727281Z  INFO store:get_block_inputs:select_block_header_by_block_num: miden-store: crates/store/src/db/mod.rs:249: close, time.busy: 22.5µs, time.idle: 201µs
2024-09-06T16:35:45.727317Z  INFO store:get_block_inputs:select_note_inclusion_proofs: miden-store: crates/store/src/db/mod.rs:370: new
2024-09-06T16:35:45.727395Z  INFO store:get_block_inputs:select_note_inclusion_proofs: miden-store: crates/store/src/db/mod.rs:370: close, time.busy: 12.6µs, time.idle: 65.4µs
2024-09-06T16:35:45.727409Z  INFO store:get_block_inputs:select_block_headers: miden-store: crates/store/src/db/mod.rs:265: new
2024-09-06T16:35:45.727493Z  INFO store:get_block_inputs:select_block_headers: miden-store: crates/store/src/db/mod.rs:265: close, time.busy: 12.3µs, time.idle: 72.2µs
2024-09-06T16:35:45.727517Z  INFO store:get_block_inputs: miden-store: crates/store/src/server/api.rs:383: close, time.busy: 191µs, time.idle: 322µs


Then it just stops responding and no more logs are printed
```

To add a little context after doing some tests:

- Running any transaction seems to stall the node indefinitely. Running the CLI for example and minting a note crashes the node as well.
- There seems to be some recursion problem. If you run the node in testing (@tomyrd was running it on release), you get a stack overflow:

```
2024-09-06T17:06:53.128378Z  INFO store:get_block_inputs: miden-store: crates/store/src/server/api.rs:383: close, time.busy: 1.76ms, time.idle: 3.39ms

thread 'tokio-runtime-worker' has overflowed its stack
fatal runtime error: stack overflow
/bin/sh: line 1: 51404 Abort trap: 6           cargo run --bin miden-node --features "testing" -- start --config ../tests/config/miden-node.toml node
make: *** [start-node] Error 134
```

--- a/crates/proto/src/domain/digest.rs
+++ b/crates/proto/src/domain/digest.rs
@@ -187,7 +187,7 @@ impl TryFrom<&digest::Digest> for [Felt; 4] {
     type Error = ConversionError;
 
     fn try_from(value: &digest::Digest) -> Result<Self, Self::Error> {
-        value.try_into()
+        (*value).try_into()
     }
 }
 
@@ -195,7 +195,7 @@ impl TryFrom<&digest::Digest> for Digest {
     type Error = ConversionError;
 
     fn try_from(value: &digest::Digest) -> Result<Self, Self::Error> {
-        value.try_into()
+        (*value).try_into()
     }
 }
 

--- a/crates/proto/src/domain/digest.rs
+++ b/crates/proto/src/domain/digest.rs
@@ -3,7 +3,7 @@ use std::fmt::{Debug, Display, Formatter};
 use hex::{FromHex, ToHex};
 use miden_objects::{notes::NoteId, Digest, Felt, StarkField};
 
-use crate::{errors::ConversionError, generated::digest};
+use crate::{errors::ConversionError, generated::digest as proto};
 
 // CONSTANTS
 // ================================================================================================
@@ -13,19 +13,19 @@ pub const DIGEST_DATA_SIZE: usize = 32;
 // FORMATTING
 // ================================================================================================
 
-impl Display for digest::Digest {
+impl Display for proto::Digest {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
         f.write_str(&self.encode_hex::<String>())
     }
 }
 
-impl Debug for digest::Digest {
+impl Debug for proto::Digest {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
         Display::fmt(self, f)
     }
 }
 
-impl ToHex for &digest::Digest {
+impl ToHex for &proto::Digest {
     fn encode_hex<T: FromIterator<char>>(&self) -> T {
         (*self).encode_hex()
     }
@@ -35,7 +35,7 @@ impl ToHex for &digest::Digest {
     }
 }
 
-impl ToHex for digest::Digest {
+impl ToHex for proto::Digest {
     fn encode_hex<T: FromIterator<char>>(&self) -> T {
         let mut data: Vec<char> = Vec::with_capacity(DIGEST_DATA_SIZE);
         data.extend(format!("{:016x}", self.d0).chars());
@@ -55,7 +55,7 @@ impl ToHex for digest::Digest {
     }
 }
 
-impl FromHex for digest::Digest {
+impl FromHex for proto::Digest {
     type Error = ConversionError;
 
     fn from_hex<T: AsRef<[u8]>>(hex: T) -> Result<Self, Self::Error> {
@@ -74,7 +74,7 @@ impl FromHex for digest::Digest {
                 let d2 = u64::from_be_bytes(data[16..24].try_into().unwrap());
                 let d3 = u64::from_be_bytes(data[24..32].try_into().unwrap());
 
-                Ok(digest::Digest { d0, d1, d2, d3 })
+                Ok(proto::Digest { d0, d1, d2, d3 })
             },
         }
     }
@@ -83,7 +83,7 @@ impl FromHex for digest::Digest {
 // INTO
 // ================================================================================================
 
-impl From<[u64; 4]> for digest::Digest {
+impl From<[u64; 4]> for proto::Digest {
     fn from(value: [u64; 4]) -> Self {
         Self {
             d0: value[0],
@@ -94,13 +94,13 @@ impl From<[u64; 4]> for digest::Digest {
     }
 }
 
-impl From<&[u64; 4]> for digest::Digest {
+impl From<&[u64; 4]> for proto::Digest {
     fn from(value: &[u64; 4]) -> Self {
         (*value).into()
     }
 }
 
-impl From<[Felt; 4]> for digest::Digest {
+impl From<[Felt; 4]> for proto::Digest {
     fn from(value: [Felt; 4]) -> Self {
         Self {
             d0: value[0].as_int(),
@@ -111,13 +111,13 @@ impl From<[Felt; 4]> for digest::Digest {
     }
 }
 
-impl From<&[Felt; 4]> for digest::Digest {
+impl From<&[Felt; 4]> for proto::Digest {
     fn from(value: &[Felt; 4]) -> Self {
         (*value).into()
     }
 }
 
-impl From<Digest> for digest::Digest {
+impl From<Digest> for proto::Digest {
     fn from(value: Digest) -> Self {
         Self {
             d0: value[0].as_int(),
@@ -128,19 +128,19 @@ impl From<Digest> for digest::Digest {
     }
 }
 
-impl From<&Digest> for digest::Digest {
+impl From<&Digest> for proto::Digest {
     fn from(value: &Digest) -> Self {
         (*value).into()
     }
 }
 
-impl From<&NoteId> for digest::Digest {
+impl From<&NoteId> for proto::Digest {
     fn from(value: &NoteId) -> Self {
         (*value).inner().into()
     }
 }
 
-impl From<NoteId> for digest::Digest {
+impl From<NoteId> for proto::Digest {
     fn from(value: NoteId) -> Self {
         value.inner().into()
     }
@@ -149,16 +149,16 @@ impl From<NoteId> for digest::Digest {
 // FROM DIGEST
 // ================================================================================================
 
-impl From<digest::Digest> for [u64; 4] {
-    fn from(value: digest::Digest) -> Self {
+impl From<proto::Digest> for [u64; 4] {
+    fn from(value: proto::Digest) -> Self {
         [value.d0, value.d1, value.d2, value.d3]
     }
 }
 
-impl TryFrom<digest::Digest> for [Felt; 4] {
+impl TryFrom<proto::Digest> for [Felt; 4] {
     type Error = ConversionError;
 
-    fn try_from(value: digest::Digest) -> Result<Self, Self::Error> {
+    fn try_from(value: proto::Digest) -> Result<Self, Self::Error> {
         if ![value.d0, value.d1, value.d2, value.d3]
             .iter()
             .all(|v| *v < <Felt as StarkField>::MODULUS)
@@ -175,26 +175,26 @@ impl TryFrom<digest::Digest> for [Felt; 4] {
     }
 }
 
-impl TryFrom<digest::Digest> for Digest {
+impl TryFrom<proto::Digest> for Digest {
     type Error = ConversionError;
 
-    fn try_from(value: digest::Digest) -> Result<Self, Self::Error> {
+    fn try_from(value: proto::Digest) -> Result<Self, Self::Error> {
         Ok(Self::new(value.try_into()?))
     }
 }
 
-impl TryFrom<&digest::Digest> for [Felt; 4] {
+impl TryFrom<&proto::Digest> for [Felt; 4] {
     type Error = ConversionError;
 
-    fn try_from(value: &digest::Digest) -> Result<Self, Self::Error> {
+    fn try_from(value: &proto::Digest) -> Result<Self, Self::Error> {
         (*value).try_into()
     }
 }
 
-impl TryFrom<&digest::Digest> for Digest {
+impl TryFrom<&proto::Digest> for Digest {
     type Error = ConversionError;
 
-    fn try_from(value: &digest::Digest) -> Result<Self, Self::Error> {
+    fn try_from(value: &proto::Digest) -> Result<Self, Self::Error> {
         (*value).try_into()
     }
 }

